### PR TITLE
chore(dependencies): Typescript 4.1.3 [WEB-1253]

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

Typescript 4.1.x introduces some extremely useful features for us that should improve our ability to make the generated types from `gamut-system` more performant.
- Codecademy: https://github.com/codecademy-engineering/Codecademy/pull/21549

This enables support for Typescript 4.1.3.

**Dependencies**
- `prettier`: 2.1.2 => 2.2.1 
- `react-use`: 15.3.4 => 15.3.8 
- `typescript-eslint`: 4.4.1 => 4.14.1

**Storybook**
- Fixes storybook to use our own babel config and docgen options.
- Fixed tsconfig to pick up the correct files.

### PR Checklist

- [ ] Related to designs:
- [x] Related to JIRA ticket: WEB-1253
- [x] I have run this code to verify it works
- [ ] This PR includes unit tests for the code change

### Testing
1. Deploy locally
2. Run `yarn`
3. Open VSCode (after restarting it) and a component file.

Add: 
```tsx
type Props = {
  [Key in 'Click' | 'Change' as `on${ Key }` ]:  () => void
}
```
4. Save the file and see that prettier formats it correctly
5. Add the type to an existing component.
6. Run `yarn start` and see that storybook builds.
7. Check the components story page to see that it has the correct types in the props table.

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.6-alpha.3a7e80.0<br/>@codecademy/gamut-labs@7.10.6-alpha.3a7e80.0<br/>@codecademy/gamut@19.11.2-alpha.3a7e80.0<br/>@codecademy/styleguide@21.15.4-alpha.3a7e80.0<br/>@codecademy/tracking@0.4.1-alpha.3a7e80.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://600a21b096b25625cbcf62d7--gamut-preview.netlify.app>https://600a21b096b25625cbcf62d7--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/600a21b096b25625cbcf62d7>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.10-alpha.bf9c5d.0<br/>@codecademy/gamut-labs@7.13.5-alpha.bf9c5d.0<br/>@codecademy/gamut@19.11.6-alpha.bf9c5d.0<br/>@codecademy/styleguide@21.20.3-alpha.bf9c5d.0<br/>@codecademy/tracking@0.4.1-alpha.bf9c5d.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://600f91a2c12988717c08c98b--gamut-preview.netlify.app>https://600f91a2c12988717c08c98b--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/600f91a2c12988717c08c98b>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.18-alpha.648934.0<br/>@codecademy/gamut-labs@8.0.2-alpha.648934.0<br/>@codecademy/gamut@20.0.5-alpha.648934.0<br/>@codecademy/styleguide@23.0.2-alpha.648934.0<br/>@codecademy/tracking@0.4.1-alpha.648934.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://601454505b26a665a6895504--gamut-preview.netlify.app>https://601454505b26a665a6895504--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/601454505b26a665a6895504>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.6-alpha.446969.0<br/>@codecademy/gamut-icons@3.14.1-alpha.446969.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.446969.0<br/>@codecademy/gamut-labs@8.0.3-alpha.446969.0<br/>@codecademy/gamut@20.0.6-alpha.446969.0<br/>@codecademy/prettier-config@0.1.10-alpha.446969.0<br/>@codecademy/styleguide@23.1.1-alpha.446969.0<br/>@codecademy/tracking@0.4.1-alpha.446969.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://601781b534cec912a9e08ccd--gamut-preview.netlify.app>https://601781b534cec912a9e08ccd--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/601781b534cec912a9e08ccd>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.6-alpha.c1e9f7.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.c1e9f7.0<br/>@codecademy/gamut-labs@8.0.3-alpha.c1e9f7.0<br/>@codecademy/gamut@20.0.6-alpha.c1e9f7.0<br/>@codecademy/prettier-config@0.1.10-alpha.c1e9f7.0<br/>@codecademy/styleguide@23.1.1-alpha.c1e9f7.0<br/>@codecademy/tracking@0.4.1-alpha.c1e9f7.0<br/>@codecademy/webpack-config@5.2.2-alpha.c1e9f7.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.6-alpha.8c331d.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.8c331d.0<br/>@codecademy/gamut-labs@8.0.3-alpha.8c331d.0<br/>@codecademy/gamut@20.0.6-alpha.8c331d.0<br/>@codecademy/prettier-config@0.1.10-alpha.8c331d.0<br/>@codecademy/styleguide@23.1.1-alpha.8c331d.0<br/>@codecademy/tracking@0.4.1-alpha.8c331d.0<br/>@codecademy/webpack-config@5.2.2-alpha.8c331d.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.6-alpha.f9e836.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.f9e836.0<br/>@codecademy/gamut-labs@8.0.3-alpha.f9e836.0<br/>@codecademy/gamut@20.0.6-alpha.f9e836.0<br/>@codecademy/prettier-config@0.1.10-alpha.f9e836.0<br/>@codecademy/styleguide@23.1.1-alpha.f9e836.0<br/>@codecademy/tracking@0.4.1-alpha.f9e836.0<br/>@codecademy/webpack-config@5.2.2-alpha.f9e836.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60178e460d9f512bcfaa2a3e--gamut-preview.netlify.app>https://60178e460d9f512bcfaa2a3e--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60178e460d9f512bcfaa2a3e>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.6-alpha.4a9d3e.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.4a9d3e.0<br/>@codecademy/gamut-labs@8.0.3-alpha.4a9d3e.0<br/>@codecademy/gamut@20.0.6-alpha.4a9d3e.0<br/>@codecademy/prettier-config@0.1.10-alpha.4a9d3e.0<br/>@codecademy/styleguide@23.1.1-alpha.4a9d3e.0<br/>@codecademy/tracking@0.4.1-alpha.4a9d3e.0<br/>@codecademy/webpack-config@5.2.2-alpha.4a9d3e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6017bab4d3430b6224c875a2--gamut-preview.netlify.app>https://6017bab4d3430b6224c875a2--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6017bab4d3430b6224c875a2>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.729490.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.729490.0<br/>@codecademy/gamut-labs@8.1.1-alpha.729490.0<br/>@codecademy/gamut@20.0.6-alpha.729490.0<br/>@codecademy/prettier-config@0.1.10-alpha.729490.0<br/>@codecademy/styleguide@23.2.1-alpha.729490.0<br/>@codecademy/tracking@0.4.1-alpha.729490.0<br/>@codecademy/webpack-config@5.2.2-alpha.729490.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60186c1e55c7100bed1b6473--gamut-preview.netlify.app>https://60186c1e55c7100bed1b6473--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60186c1e55c7100bed1b6473>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.3d7bc4.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.3d7bc4.0<br/>@codecademy/gamut-labs@8.1.1-alpha.3d7bc4.0<br/>@codecademy/gamut@20.0.6-alpha.3d7bc4.0<br/>@codecademy/prettier-config@0.1.10-alpha.3d7bc4.0<br/>@codecademy/styleguide@23.2.1-alpha.3d7bc4.0<br/>@codecademy/tracking@0.4.1-alpha.3d7bc4.0<br/>@codecademy/webpack-config@5.2.2-alpha.3d7bc4.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6018720cce139e0df6c0fc6f--gamut-preview.netlify.app>https://6018720cce139e0df6c0fc6f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6018720cce139e0df6c0fc6f>Deploy Logs</a></p>

File: packages/gamut/src/Button/CTAButton.tsx [hunk 1]
```diff
@@ -43,9 +43,9 @@ const CTAButtonOuter = styled(ButtonOutline)`
   padding: 1px 1px 5px 5px;
 `;
 
-export const CTAButton: React.FC<React.ComponentProps<
-  typeof CTAButtonOuter
->> = ({ children, mode, ...props }) => {
+export const CTAButton: React.FC<
+  React.ComponentProps<typeof CTAButtonOuter>
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Button/CTAButton.tsx:47]: Oh I do like this change... +1 to parser/prettier fixes heh

**Review by JoshuaKGoldberg — APPROVED** [review]
Great!

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.0fe6fc.0<br/>@codecademy/gamut-illustrations@0.4.19-alpha.0fe6fc.0<br/>@codecademy/gamut-labs@8.1.1-alpha.0fe6fc.0<br/>@codecademy/gamut-styles@7.7.1-alpha.0fe6fc.0<br/>@codecademy/gamut-system@0.2.2-alpha.0fe6fc.0<br/>@codecademy/gamut@20.0.6-alpha.0fe6fc.0<br/>@codecademy/markdown-overrides@0.4.5-alpha.0fe6fc.0<br/>@codecademy/prettier-config@0.1.10-alpha.0fe6fc.0<br/>@codecademy/styleguide@23.2.1-alpha.0fe6fc.0<br/>@codecademy/tracking@0.4.1-alpha.0fe6fc.0<br/>@codecademy/webpack-config@5.2.2-alpha.0fe6fc.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://601873899bf1bf11ef6a82f6--gamut-preview.netlify.app>https://601873899bf1bf11ef6a82f6--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/601873899bf1bf11ef6a82f6>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
love it

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.f9f290.0<br/>@codecademy/gamut-illustrations@0.4.23-alpha.f9f290.0<br/>@codecademy/gamut-labs@8.1.6-alpha.f9f290.0<br/>@codecademy/gamut-styles@7.7.2-alpha.f9f290.0<br/>@codecademy/gamut-system@0.2.3-alpha.f9f290.0<br/>@codecademy/gamut@21.1.2-alpha.f9f290.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.f9f290.0<br/>@codecademy/prettier-config@0.1.10-alpha.f9f290.0<br/>@codecademy/styleguide@24.0.4-alpha.f9f290.0<br/>@codecademy/tracking@0.4.2-alpha.f9f290.0<br/>@codecademy/webpack-config@5.2.3-alpha.f9f290.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6019b3c0b355d60e06acc243--gamut-preview.netlify.app>https://6019b3c0b355d60e06acc243--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6019b3c0b355d60e06acc243>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.68effc.0<br/>@codecademy/gamut-illustrations@0.4.23-alpha.68effc.0<br/>@codecademy/gamut-labs@8.1.6-alpha.68effc.0<br/>@codecademy/gamut-styles@7.7.2-alpha.68effc.0<br/>@codecademy/gamut-system@0.2.3-alpha.68effc.0<br/>@codecademy/gamut@21.1.2-alpha.68effc.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.68effc.0<br/>@codecademy/prettier-config@0.1.10-alpha.68effc.0<br/>@codecademy/styleguide@24.0.4-alpha.68effc.0<br/>@codecademy/tracking@0.4.2-alpha.68effc.0<br/>@codecademy/webpack-config@5.2.3-alpha.68effc.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6019b5cc9ce885111cfe5121--gamut-preview.netlify.app>https://6019b5cc9ce885111cfe5121--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6019b5cc9ce885111cfe5121>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.ac0155.0<br/>@codecademy/gamut-illustrations@0.4.23-alpha.ac0155.0<br/>@codecademy/gamut-labs@8.1.6-alpha.ac0155.0<br/>@codecademy/gamut-styles@7.7.2-alpha.ac0155.0<br/>@codecademy/gamut-system@0.2.3-alpha.ac0155.0<br/>@codecademy/gamut@21.1.2-alpha.ac0155.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.ac0155.0<br/>@codecademy/prettier-config@0.1.10-alpha.ac0155.0<br/>@codecademy/styleguide@24.0.4-alpha.ac0155.0<br/>@codecademy/tracking@0.4.2-alpha.ac0155.0<br/>@codecademy/webpack-config@5.2.3-alpha.ac0155.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6019c0c5de83bd1d59965e50--gamut-preview.netlify.app>https://6019c0c5de83bd1d59965e50--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6019c0c5de83bd1d59965e50>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.c673b3.0<br/>@codecademy/gamut-illustrations@0.4.24-alpha.c673b3.0<br/>@codecademy/gamut-labs@8.2.1-alpha.c673b3.0<br/>@codecademy/gamut-styles@7.7.2-alpha.c673b3.0<br/>@codecademy/gamut-system@0.2.3-alpha.c673b3.0<br/>@codecademy/gamut@21.2.1-alpha.c673b3.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.c673b3.0<br/>@codecademy/prettier-config@0.1.10-alpha.c673b3.0<br/>@codecademy/styleguide@24.2.1-alpha.c673b3.0<br/>@codecademy/tracking@0.4.2-alpha.c673b3.0<br/>@codecademy/webpack-config@5.2.3-alpha.c673b3.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6019d272ce9ffb00990a55d6--gamut-preview.netlify.app>https://6019d272ce9ffb00990a55d6--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6019d272ce9ffb00990a55d6>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.c242f6.0<br/>@codecademy/gamut-illustrations@0.4.24-alpha.c242f6.0<br/>@codecademy/gamut-labs@8.3.1-alpha.c242f6.0<br/>@codecademy/gamut-styles@7.7.2-alpha.c242f6.0<br/>@codecademy/gamut-system@0.2.3-alpha.c242f6.0<br/>@codecademy/gamut@21.2.1-alpha.c242f6.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.c242f6.0<br/>@codecademy/prettier-config@0.1.10-alpha.c242f6.0<br/>@codecademy/styleguide@24.2.3-alpha.c242f6.0<br/>@codecademy/tracking@0.4.2-alpha.c242f6.0<br/>@codecademy/webpack-config@5.2.3-alpha.c242f6.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://601abb1097e632009adb3216--gamut-preview.netlify.app>https://601abb1097e632009adb3216--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/601abb1097e632009adb3216>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/eslint-config@3.1.7-alpha.7da8db.0<br/>@codecademy/gamut-illustrations@0.4.26-alpha.7da8db.0<br/>@codecademy/gamut-labs@8.3.3-alpha.7da8db.0<br/>@codecademy/gamut-styles@7.7.2-alpha.7da8db.0<br/>@codecademy/gamut-system@0.2.3-alpha.7da8db.0<br/>@codecademy/gamut@21.2.3-alpha.7da8db.0<br/>@codecademy/markdown-overrides@0.4.6-alpha.7da8db.0<br/>@codecademy/prettier-config@0.1.10-alpha.7da8db.0<br/>@codecademy/styleguide@24.2.5-alpha.7da8db.0<br/>@codecademy/tracking@0.4.2-alpha.7da8db.0<br/>@codecademy/webpack-config@5.2.3-alpha.7da8db.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://601c1518b58ba36e42a4062e--gamut-preview.netlify.app>https://601c1518b58ba36e42a4062e--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/601c1518b58ba36e42a4062e>Deploy Logs</a></p>

**Pull request merged**